### PR TITLE
fixes multi-namespace bucket class page, table and dropdown sync

### PR DESCRIPTION
**Pull request by vbnrh**
Fixes https://issues.redhat.com/browse/RHSTOR-1791

Signed-off-by: Vineet Badrinath <vbadrina@redhat.com>

Screenshot from 2021-04-15 13-16-28
Screenshot from 2021-04-15 13-16-06
Screenshot from 2021-04-15 13-16-02

### CODE REVIEW

**Comment by afreen23** [comment]
Please add screenhots as well.

**Comment by vbnrh** [comment]
> Please add screenhots as well.

done

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx [hunk 1]
```diff
@@ -23,22 +23,26 @@ export const MultiNamespaceStorePage: React.FC<MultiNamespaceStoreProps> = React
...
     const [selectedCount, setSelectedCount] = React.useState(state.readNamespaceStore.length);
     const [enabledItems, setEnabledItems] = React.useState([]);
+    const [selectedItem, setSelectedItem] = React.useState('');
 
     React.useEffect(() => {
       const readItems = state.readNamespaceStore.map(getName);
       setEnabledItems(readItems);
-    }, [state.readNamespaceStore]);
+      setSelectedItem(getName(state.writeNamespaceStore[0]) || '');
+    }, [state.readNamespaceStore, state.writeNamespaceStore, selectedItem]);
```
Comment (GowthamShanmugam) [inline frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx:32]: Line 45, is already updating selectedItem, you are overriding same value multiple times

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx [hunk 2]
```diff
@@ -23,22 +23,26 @@ export const MultiNamespaceStorePage: React.FC<MultiNamespaceStoreProps> = React
...
       setSelectedCount(selectedNamespaceStore.length);
       if (!selectedNamespaceStore.map(getName).includes(getName(state.writeNamespaceStore[0]))) {
         dispatch({ type: 'setWriteNamespaceStore', value: [] });
+        setSelectedItem('');
       }
     };
 
     const onSelectNamespaceStoreDropdown = (selectedNamespaceStore: NamespaceStoreKind) => {
       dispatch({ type: 'setWriteNamespaceStore', value: [selectedNamespaceStore] });
+      setSelectedItem(getName(state.writeNamespaceStore[0]));
```
Comment (GowthamShanmugam) [inline frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx:45]: dispatch is an async update so state.writeNamespaceStore[0] wont gives you the latest value

try `setSelectedItem(getName(selectedNamespaceStore));`

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx [hunk 3]
```diff
@@ -23,22 +23,26 @@ export const MultiNamespaceStorePage: React.FC<MultiNamespaceStoreProps> = React
     const openModal = () => NamespaceStoreModal({ namespace });
     const [selectedCount, setSelectedCount] = React.useState(state.readNamespaceStore.length);
     const [enabledItems, setEnabledItems] = React.useState([]);
+    const [selectedItem, setSelectedItem] = React.useState('');
```
Comment (GowthamShanmugam) [inline frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx:26]: React.useState<string>('');

Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx]: Do you mean React.useState(''"); with double quotes..... It already is React.useState('');(single quotes)

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx (hunk 1 above)
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx:33]: this is needed in edit-bucket class where we need to immediately populate the fields on the very first render. I can do it here or above. I've moved it above now

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx (hunk 2 above)
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx]: makes sense ! thanks

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx (hunk 3 above)
Comment (GowthamShanmugam) [reply frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx]: Ah, i mean to say use React.useState<string>('');

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx [hunk 4]
```diff
@@ -23,22 +23,27 @@ export const MultiNamespaceStorePage: React.FC<MultiNamespaceStoreProps> = React
     const openModal = () => NamespaceStoreModal({ namespace });
     const [selectedCount, setSelectedCount] = React.useState(state.readNamespaceStore.length);
     const [enabledItems, setEnabledItems] = React.useState([]);
+    const [selectedItem, setSelectedItem] = React.useState(
+      getName(state.writeNamespaceStore[0]) ? getName(state.writeNamespaceStore[0]) : '',
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx:27]:
```suggestion
      getName(state.writeNamespaceStore[0]) ?? '',
```

File: frontend/packages/ceph-storage-plugin/src/components/namespace-store/namespace-store-dropdown.tsx [hunk 5]
```diff
@@ -117,7 +111,7 @@ export const NamespaceStoreDropdown: React.FC<NamespaceStoreDropdownProps> = ({
             data-test="nns-dropdown-toggle"
             onToggle={() => setOpen(!isOpen)}
           >
-            {selectedKey || nsName || t('ceph-storage-plugin~Select a namespace store')}
+            {selectedKey || t('ceph-storage-plugin~Select a namespace store')}
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/namespace-store/namespace-store-dropdown.tsx:114]:
```suggestion
            {selectedKey || t('ceph-storage-plugin~Select a NamespaceStore')}
```
isn't it?

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx (hunk 3 above)
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx]: No problem :-) Lets keep it the same... typescript will automatically infer the type, no need to specify it

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx [hunk 6]
```diff
@@ -23,22 +23,27 @@ export const MultiNamespaceStorePage: React.FC<MultiNamespaceStoreProps> = React
...
+    );
 
     React.useEffect(() => {
       const readItems = state.readNamespaceStore.map(getName);
       setEnabledItems(readItems);
-    }, [state.readNamespaceStore]);
+    }, [state.readNamespaceStore, state.writeNamespaceStore, selectedItem]);
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx:33]:
```suggestion
    }, [state.readNamespaceStore]);
```
Why do we need to watch on writeNamespaceStore and selectedItem?
If that changes readNamespaceStore somehow then wouldn't readNamespaceStore just get retriggered?

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx (hunk 4 above)
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx]: fixed

File: frontend/packages/ceph-storage-plugin/src/components/namespace-store/namespace-store-dropdown.tsx (hunk 5 above)
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/namespace-store/namespace-store-dropdown.tsx:114]: we are using the same style in backing store dropdown. Lets keep it same. I am planning to make the component generic.

File: frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx (hunk 6 above)
Comment (vbnrh) [reply frontend/packages/ceph-storage-plugin/src/components/bucket-class/wizard-pages/namespace-store-pages/multi-namespace-store.tsx]: true. forgot to remove it , thanks

**Pull request merged**